File: src/main/java/io/github/aalsanie/codes/Constraints.java
package io.github.aalsanie.codes;

final class Constraints {
    private static final int MAX_NAMESPACE_LENGTH = 128;
    private static final int MAX_NAME_LENGTH = 64;
    private static final int MAX_MESSAGE_LENGTH = 1024;
    private static final int MAX_DETAIL_LENGTH = 4096;
    private static final int MAX_PATH_LENGTH = 256;

    private Constraints() {
    }

    static void requireNamespace(String namespace) {
        if (namespace == null) {
            throw new NullPointerException("namespace");
        }
        if (namespace.isEmpty() || namespace.length() > MAX_NAMESPACE_LENGTH) {
            throw new IllegalArgumentException(
                "namespace length must be between 1 and " + MAX_NAMESPACE_LENGTH
            );
        }

        String[] segments = namespace.split("\\.", -1);
        for (String segment : segments) {
            if (segment.isEmpty()) {
                throw new IllegalArgumentException("namespace must not contain empty segments");
            }
            if (segment.length() > 63) {
                throw new IllegalArgumentException("namespace segment length must not exceed 63");
            }
            if (!isLowerAsciiLetter(segment.charAt(0))) {
                throw new IllegalArgumentException(
                    "namespace segments must start with a lowercase ASCII letter"
                );
            }
            char last = segment.charAt(segment.length() - 1);
            if (!(isLowerAsciiLetter(last) || isAsciiDigit(last))) {
                throw new IllegalArgumentException(
                    "namespace segments must end with a lowercase ASCII letter or digit"
                );
            }
            for (int i = 0; i < segment.length(); i++) {
                char c = segment.charAt(i);
                if (!(isLowerAsciiLetter(c) || isAsciiDigit(c) || c == '-')) {
                    throw new IllegalArgumentException(
                        "namespace segments may contain only lowercase ASCII letters, digits, and hyphens"
                    );
                }
            }
        }
    }

    static void requireName(String name) {
        if (name == null) {
            throw new NullPointerException("name");
        }
        if (name.isEmpty() || name.length() > MAX_NAME_LENGTH) {
            throw new IllegalArgumentException(
                "name length must be between 1 and " + MAX_NAME_LENGTH
            );
        }
        if (!isUpperAsciiLetter(name.charAt(0))) {
            throw new IllegalArgumentException("name must start with an uppercase ASCII letter");
        }
        for (int i = 0; i < name.length(); i++) {
            char c = name.charAt(i);
            if (!(isUpperAsciiLetter(c) || isAsciiDigit(c) || c == '_')) {
                throw new IllegalArgumentException(
                    "name may contain only uppercase ASCII letters, digits, and underscores"
                );
            }
        }
    }

    static void requireMessage(String message) {
        requireText(message, "message", MAX_MESSAGE_LENGTH);
    }

    static void requireDetail(String detail) {
        requireText(detail, "detail", MAX_DETAIL_LENGTH);
    }

    static void requirePath(String path) {
        requireText(path, "path", MAX_PATH_LENGTH);
    }

    private static void requireText(String value, String label, int maxLength) {
        if (value == null) {
            throw new NullPointerException(label);
        }
        if (value.isBlank()) {
            throw new IllegalArgumentException(label + " must not be blank");
        }
        if (value.length() > maxLength) {
            throw new IllegalArgumentException(label + " length must not exceed " + maxLength);
        }
        if (value.indexOf('\0') >= 0) {
            throw new IllegalArgumentException(label + " must not contain NUL");
        }
    }

    private static boolean isLowerAsciiLetter(char c) {
        return c >= 'a' && c <= 'z';
    }

    private static boolean isUpperAsciiLetter(char c) {
        return c >= 'A' && c <= 'Z';
    }

    private static boolean isAsciiDigit(char c) {
        return c >= '0' && c <= '9';
    }
}

File: src/main/kotlin/io/github/aalsanie/codes/Constraints.kt
package io.github.aalsanie.codes

internal object Constraints {
    const val MAX_NAMESPACE_LENGTH = 128
    const val MAX_NAME_LENGTH = 64
    const val MAX_MESSAGE_LENGTH = 1024
    const val MAX_DETAIL_LENGTH = 4096
    const val MAX_PATH_LENGTH = 256

    fun requireNamespace(namespace: String) {
        require(namespace.length in 1..MAX_NAMESPACE_LENGTH) {
            "namespace length must be between 1 and $MAX_NAMESPACE_LENGTH"
        }

        val segments = namespace.split('.')
        require(segments.none { it.isEmpty() }) {
            "namespace must not contain empty segments"
        }

        segments.forEach { segment ->
            require(segment.length <= 63) {
                "namespace segment length must not exceed 63"
            }
            require(segment.first() in 'a'..'z') {
                "namespace segments must start with a lowercase ASCII letter"
            }
            require(segment.last() in 'a'..'z' || segment.last() in '0'..'9') {
                "namespace segments must end with a lowercase ASCII letter or digit"
            }
            require(segment.all { it in 'a'..'z' || it in '0'..'9' || it == '-' }) {
                "namespace segments may contain only lowercase ASCII letters, digits, and hyphens"
            }
        }
    }

    fun requireName(name: String) {
        require(name.length in 1..MAX_NAME_LENGTH) {
            "name length must be between 1 and $MAX_NAME_LENGTH"
        }
        require(name.first() in 'A'..'Z') {
            "name must start with an uppercase ASCII letter"
        }
        require(name.all { it in 'A'..'Z' || it in '0'..'9' || it == '_' }) {
            "name may contain only uppercase ASCII letters, digits, and underscores"
        }
    }

    fun requireMessage(message: String) {
        require(message.isNotBlank()) {
            "message must not be blank"
        }
        require(message.length <= MAX_MESSAGE_LENGTH) {
            "message length must not exceed $MAX_MESSAGE_LENGTH"
        }
        require('\u0000' !in message) {
            "message must not contain NUL"
        }
    }

    fun requireDetail(detail: String) {
        require(detail.isNotBlank()) {
            "detail must not be blank"
        }
        require(detail.length <= MAX_DETAIL_LENGTH) {
            "detail length must not exceed $MAX_DETAIL_LENGTH"
        }
        require('\u0000' !in detail) {
            "detail must not contain NUL"
        }
    }

    fun requirePath(path: String) {
        require(path.isNotBlank()) {
            "path must not be blank"
        }
        require(path.length <= MAX_PATH_LENGTH) {
            "path length must not exceed $MAX_PATH_LENGTH"
        }
        require('\u0000' !in path) {
            "path must not contain NUL"
        }
    }
}
